Chat App/ViewModel/Pub/PubRegisterViewModel.swift

```diff
@@ -0,0 +1,154 @@
+//
+//  PubRegisterViewModel.swift
+//  Chat App
+//
+//  Created by Amiel Jireh Cordova on 12/14/20.
+//  Copyright © 2020 Cordova, Jireh. All rights reserved.
+//
+
+import Foundation
+import RxCocoa
+import RxSwift
+
+protocol RegisterViewModelInputs {
+    func verifyUsernameAvailability(userInput: String)
+    func verifyUserInput(userInput: String) -> Bool
+    func verifyPasswordMatch(userInput: String!, comparable: String!) -> Bool
+    func registerUser(submittedUsername: String!, password: String!)
+}
+
+protocol RegisterViewModelOutputs {
+    var alertTitle: String { get }
+    var alertMessage: String { get }
+    var alertActionLabel: String { get }
+    var shouldShowAlert: PublishSubject<Bool> { get }
+    var shouldShowLoading: PublishSubject<Bool> { get }
+    var isUsernameAvailable: PublishSubject<Bool> { get }
+    var shouldProceedtoServer: PublishSubject<Bool> { get }
+}
+
+protocol RegisterViewModelTypes {
+    var inputs: RegisterViewModelInputs { get }
+    var outputs: RegisterViewModelOutputs { get }
+}
+
+class PubRegisterViewModel: RegisterViewModelInputs, RegisterViewModelOutputs, RegisterViewModelTypes {
+    var shouldShowAlert = PublishSubject<Bool>()
+    var shouldShowLoading = PublishSubject<Bool>()
+    var isUsernameAvailable = PublishSubject<Bool>()
+    var shouldProceedtoServer = PublishSubject<Bool>()
+    var inputs: RegisterViewModelInputs { return self }
+    var outputs: RegisterViewModelOutputs { return self }
+    
+    var alertTitle = ""
+    var alertMessage = ""
+    var alertActionLabel = ""
+    var userManager: UserManagementProtocol
+    
+    private let disposeBag = DisposeBag()
+    
+    init() {
+        self.userManager = UserManagementService()
+        setUpObservers()
+    }
+    
+    /**
+     Verifies that user input meets character limit requirements.
+     - Parameter userInput: submitted user input.
+     - Returns a Bool value that indicates if user input was valid.
+     */
+    func verifyUserInput(userInput: String) -> Bool {
+        userInput.count >= 8 && userInput.count <= 16
+    }
+    
+    /**
+     Verifies that the inputted username is available for registration.
+     - Parameter userInput: submitted user input.
+     */
+    func verifyUsernameAvailability(userInput: String) {
+        userManager.checkUsernameAvailability(userInput: userInput)
+    }
+    
+    /**
+     Registers a new Pub user.
+     - Parameter submittedUsername: submitted user name.
+     - Parameter password: submitted password.
+     */
+    func registerUser(submittedUsername: String!, password: String!) {
```

Comment:
is  `!`necessary here?

Reply:
removed force cast usage.

Reply:
Refactored usage and method call. Removed conditionals.